### PR TITLE
feat: hardware wallet connection utils

### PR DESCRIPTION
**Pull request by montelaidev**
## **Description**

This pr introduces hardware wallet connection utils that can be used to check for permissions and connect to the hardware devices.

[Open in GitHub Codespaces](https://codespaces.new/MetaMask/metamask-extension/pull/39160?quickstart=1)

## **Changelog**

CHANGELOG entry: null

## **Related issues**

Related: https://consensyssoftware.atlassian.net/browse/MUL-1299?atlOrigin=eyJpIjoiZWZlYjE4M2NiOWVmNDk0N2I3MzA4MzMzZTg2M2U1YzYiLCJwIjoiaiJ9

## **Manual testing steps**

N/a. These are utils that are not currently used.

## **Screenshots/Recordings**

N/a. These are utils that are not currently used.

## **Pre-merge author checklist**

- [x] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Extension Coding Standards](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

---

> [!NOTE]
> Adds web connection utilities for hardware wallets with comprehensive tests.
> 
> - New `ui/contexts/hardware-wallets/webConnectionUtils.ts` provides availability checks (`isWebHidAvailable`, `isWebUsbAvailable`), permission checks/requests (`check*Permission`, `request*Permission`), device queries (`getConnected*Devices`, `is*Connected`, `get*DeviceId`), and event subscriptions (`subscribeToWebHidEvents`, `subscribeToWebUsbEvents`) for Ledger (WebHID) and Trezor (WebUSB)
> - Adds `TREZOR_USB_VENDOR_IDS` to `shared/constants/hardware-wallets.ts` and re-exports via `ui/contexts/hardware-wallets/index.ts`
> - Extensive unit tests in `webConnectionUtils.test.ts` covering success, denial/cancellation, and edge cases
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 6a43b9a419d4884177b57a10631e8b8dc3dbf51a. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: shared/constants/hardware-wallets.ts [hunk 1]
```diff
@@ -139,6 +139,12 @@ export enum MarketingActionNames {
 
 export const LEDGER_USB_VENDOR_ID = '0x2c97';
 
+export const TREZOR_USB_VENDOR_IDS = [
+  { vendorId: 0x534c, productId: 0x0001 }, // Trezor One
+  { vendorId: 0x1209, productId: 0x53c0 }, // Trezor Model T
+  { vendorId: 0x1209, productId: 0x53c1 }, // Trezor Safe 5
```
Comment (owencraston) [inline shared/constants/hardware-wallets.ts]: this is missing the Trezor Safe 7

File: ui/contexts/hardware-wallets/types.ts [hunk 2]
```diff
@@ -0,0 +1,56 @@
+/**
+ * Types and interfaces for hardware wallet functionality
+ */
+
+/**
+ * Hardware wallet types supported by the application
+ */
+export enum HardwareWalletType {
+  Ledger = 'ledger',
```
Comment (owencraston) [inline ui/contexts/hardware-wallets/types.ts:5]: why are these lowercase? In our analytics we defined these values as 

```
      - Ledger
      - Trezor
      - QR Hardware
      - Lattice
```

File: ui/contexts/hardware-wallets/webConnectionUtils.ts [hunk 3]
```diff
@@ -0,0 +1,446 @@
...
+      device,
+      HardwareWalletType.Trezor,
+    );
+
+    return hasTrezorDevice;
+  } catch (error) {
+    const errorMessage = error instanceof Error ? error.message : String(error);
+
+    // User cancelled the dialog
+    if (errorMessage.includes('cancelled') || errorMessage.includes('denied')) {
```
Comment (owencraston) [inline ui/contexts/hardware-wallets/webConnectionUtils.ts]: should these be constants?

File: ui/contexts/hardware-wallets/types.ts (hunk 2 above)
Comment (montelaidev) [reply ui/contexts/hardware-wallets/types.ts:5]: These values aren't used in the analytics. I made the names camelCase to match the localized key

File: ui/contexts/hardware-wallets/webConnectionUtils.ts [hunk 4]
```diff
@@ -0,0 +1,446 @@
...
+ * This will show the browser's device selection dialog
+ */
+export async function requestWebHIDPermission(): Promise<boolean> {
+  if (!isWebHIDAvailable()) {
+    return false;
+  }
+
+  try {
+    const devices = await window.navigator.hid.requestDevice({
+      filters: [{ vendorId: LEDGER_USB_VENDOR_ID as unknown as number }],
```
[the remark this answers is not in the document]
Comment (montelaidev) [reply ui/contexts/hardware-wallets/webConnectionUtils.ts]: The hex is fine.

File: ui/contexts/hardware-wallets/webConnectionUtils.ts (hunk 3 above)
Comment (montelaidev) [reply ui/contexts/hardware-wallets/webConnectionUtils.ts]: The error messages are from the webhid/webusb. These might not be standardized between browsers. I've removed this code, because it is redundant

File: shared/constants/hardware-wallets.ts (hunk 1 above)
Comment (montelaidev) [reply shared/constants/hardware-wallets.ts]: I've removed the incorrect comments. Safe 7 is still detected with the existing ids.

File: shared/constants/hardware-wallets.ts [hunk 5]
```diff
@@ -139,6 +139,12 @@ export enum MarketingActionNames {
 
 export const LEDGER_USB_VENDOR_ID = '0x2c97';
 
+export const TREZOR_USB_VENDOR_IDS = [
+  { vendorId: 0x534c, productId: 0x0001 },
+  { vendorId: 0x1209, productId: 0x53c0 },
+  { vendorId: 0x1209, productId: 0x53c1 },
+];
```
Comment (gantunesr) [inline shared/constants/hardware-wallets.ts:146]: I remember you had comments to know which ID is which device, I found it quite convenient

File: ui/contexts/hardware-wallets/webConnectionUtils.ts [hunk 6]
```diff
@@ -0,0 +1,432 @@
...
+    typeof window !== 'undefined' &&
+    typeof window.navigator !== 'undefined' &&
+    'hid' in window.navigator
+  );
+}
+
+/**
+ * Check if WebUSB is available in the current browser
+ */
+export function isWebUSBAvailable(): boolean {
```
Comment (gantunesr) [inline ui/contexts/hardware-wallets/webConnectionUtils.ts]: nit: similarly, should it be `isWebUsbAvailable`?

File: ui/contexts/hardware-wallets/webConnectionUtils.ts [hunk 7]
```diff
@@ -0,0 +1,432 @@
...
+    );
+
+    if (hasLedgerDevice) {
+      return HardwareConnectionPermissionState.Granted;
+    }
+
+    // No paired devices means we need to request permission
+    return HardwareConnectionPermissionState.Prompt;
+  } catch (error) {
+    console.error(LOG_TAG, 'Error checking WebHID permission:', error);
```
Comment (gantunesr) [inline ui/contexts/hardware-wallets/webConnectionUtils.ts]: are the `console.error`s left on purpose?

File: ui/contexts/hardware-wallets/webConnectionUtils.ts [hunk 8]
```diff
@@ -0,0 +1,432 @@
...
+ * This will show the browser's device selection dialog
+ */
+export async function requestWebHIDPermission(): Promise<boolean> {
+  if (!isWebHIDAvailable()) {
+    return false;
+  }
+
+  try {
+    const devices = await window.navigator.hid.requestDevice({
+      filters: [{ vendorId: LEDGER_USB_VENDOR_ID as unknown as number }],
```
Comment (gantunesr) [inline ui/contexts/hardware-wallets/webConnectionUtils.ts]: I see you marked the cursor comment as resolved but I don't follow, why not use `Number(LEDGER_USB_VENDOR_ID)` here too? it seems more correct even if hex could be handled in this case

File: ui/contexts/hardware-wallets/webConnectionUtils.ts
Comment (gantunesr) [inline ui/contexts/hardware-wallets/webConnectionUtils.ts:1]: nit: I see multiple methods and variable names that use `HID` and `USB`, should those be update to be `Hid` and `Usb`

Comment (gantunesr) [inline ui/contexts/hardware-wallets/webConnectionUtils.ts:1]: nit: could you add an `index.ts` for exporting the methods?

**Review by gantunesr — COMMENTED** [review]
Looks good in general! left mainly minor comments

File: ui/contexts/hardware-wallets/webConnectionUtils.ts [hunk 9]
```diff
@@ -0,0 +1,432 @@
...
+    const devices = await window.navigator.hid.getDevices();
+
+    // Check if any Ledger devices are paired
+    const hasLedgerDevice = devices.some((device) =>
+      matchesDeviceFilters(device, HardwareWalletType.Ledger),
+    );
+
+    if (hasLedgerDevice) {
+      return HardwareConnectionPermissionState.Granted;
+    }
```
Comment (ccharly) [inline ui/contexts/hardware-wallets/webConnectionUtils.ts:114]: WDYT about making it slightly more generic like:
```suggestion
/**
 * Check if WebHID permission is granted by checking for paired devices
 */
export async function checkWebHIDPermission(walletType: HardwareWalletType): Promise<HardwareConnectionPermissionState> {
  if (!isWebHIDAvailable()) {
    return HardwareConnectionPermissionState.Denied;
  }

  try {
    const devices = await window.navigator.hid.getDevices();

    // Check if any HID devices are paired
    const hasDevice = devices.some((device) =>
      matchesDeviceFilters(device, walletType),
    );

    if (hasDevice) {
      return HardwareConnectionPermissionState.Granted;
    }
```

File: ui/contexts/hardware-wallets/webConnectionUtils.ts [hunk 10]
```diff
@@ -0,0 +1,432 @@
...
+  } catch (error) {
+    console.error(LOG_TAG, 'Error checking WebHID permission:', error);
+    return HardwareConnectionPermissionState.Unknown;
+  }
+}
+
+/**
+ * Check if WebUSB permission is granted by checking for paired devices
+ */
+export async function checkWebUSBPermission(): Promise<HardwareConnectionPermissionState> {
```
Comment (ccharly) [inline ui/contexts/hardware-wallets/webConnectionUtils.ts]: We could follow the same logic here too

File: ui/contexts/hardware-wallets/webConnectionUtils.ts [hunk 11]
```diff
@@ -0,0 +1,432 @@
...
+    default:
+      return false;
+  }
+}
+
+/**
+ * Request WebHID permission from the user
+ * This will show the browser's device selection dialog
+ */
+export async function requestWebHIDPermission(): Promise<boolean> {
```
Comment (ccharly) [inline ui/contexts/hardware-wallets/webConnectionUtils.ts]: Same here, we could pass the vendor ID in the parameter, making the whole function more generic IMO

File: ui/contexts/hardware-wallets/webConnectionUtils.ts [hunk 12]
```diff
@@ -0,0 +1,432 @@
...
+  } catch {
+    return false;
+  }
+}
+
+/**
+ * Request WebUSB permission from the user
+ * This will show the browser's device selection dialog
+ */
+export async function requestWebUSBPermission(): Promise<boolean> {
```
Comment (ccharly) [inline ui/contexts/hardware-wallets/webConnectionUtils.ts]: Same here

File: ui/contexts/hardware-wallets/webConnectionUtils.ts
Comment (mathieuartu) [inline ui/contexts/hardware-wallets/webConnectionUtils.ts:1]: I think it would be best to have an index file, WDYT?

File: ui/contexts/hardware-wallets/webConnectionUtils.ts [hunk 13]
```diff
@@ -0,0 +1,432 @@
...
+}
+
+/**
+ * Check if a device matches the vendor filters for a specific hardware wallet type
+ *
+ * @param device - The device to check (HIDDevice or USBDevice)
+ * @param walletType - The hardware wallet type to match against
+ * @returns true if the device matches the filters for the specified wallet type
+ */
+function matchesDeviceFilters(
```
Comment (ccharly) [inline ui/contexts/hardware-wallets/webConnectionUtils.ts]: Cosmetic: WDYT about naming `isHardwareWalletDevice` instead?

File: ui/contexts/hardware-wallets/webConnectionUtils.ts (hunk 7 above)
Comment (montelaidev) [reply ui/contexts/hardware-wallets/webConnectionUtils.ts]: Its for debugging, it can be removed.

File: shared/constants/hardware-wallets.ts [hunk 14]
```diff
@@ -139,6 +139,12 @@ export enum MarketingActionNames {
 
 export const LEDGER_USB_VENDOR_ID = '0x2c97';
```
Comment (mathieuartu) [inline shared/constants/hardware-wallets.ts:140]: I know it's out of scope, but having this without quotes would probably solve your `unknow as number` cast in the other file

File: ui/contexts/hardware-wallets/webConnectionUtils.ts [hunk 15]
```diff
@@ -0,0 +1,432 @@
...
+  if (!isWebUSBAvailable()) {
+    return [];
+  }
+
+  try {
+    const devices = await window.navigator.usb.getDevices();
+    return devices.filter((device) =>
+      matchesDeviceFilters(device, HardwareWalletType.Trezor),
+    );
+  } catch (error) {
```
Comment (mathieuartu) [inline ui/contexts/hardware-wallets/webConnectionUtils.ts]:
```suggestion
  } catch {
```

File: ui/contexts/hardware-wallets/webConnectionUtils.ts (hunk 7 above)
Comment (mathieuartu) [reply ui/contexts/hardware-wallets/webConnectionUtils.ts]: I made a suggestion below to remove the `(error)`. But if logging is on purpose here, shouldn't we log (also using `LOG_TAG` everywhere else?)

File: shared/constants/hardware-wallets.ts (hunk 5 above)
Comment (montelaidev) [reply shared/constants/hardware-wallets.ts:146]: The comments were not completely correct, so i removed them.

File: ui/contexts/hardware-wallets/webConnectionUtils.test.ts [hunk 16]
```diff
@@ -0,0 +1,1526 @@
...
+      (window.navigator.hid.getDevices as jest.Mock).mockResolvedValue([]);
+
+      const result = await checkWebHidPermission(HardwareWalletType.Ledger);
+
+      expect(result).toBe(HardwareConnectionPermissionState.Prompt);
+      expect(window.navigator.hid.getDevices).toHaveBeenCalled();
+    });
+
+    it('returns Prompt when paired devices exist but are not Ledger', async () => {
+      (window.navigator.hid as jest.Mocked<HID>).getDevices.mockResolvedValue([
```
Comment (ccharly) [inline ui/contexts/hardware-wallets/webConnectionUtils.test.ts]: Maybe we should introduce a `getMockedHid()` and `getMockedUsb()` helper that would cast those.

That would reduce all those redundant `as jest.Mocked<...>`, WDYT?

File: ui/contexts/hardware-wallets/webConnectionUtils.test.ts [hunk 17]
```diff
@@ -0,0 +1,1526 @@
...
+
+    it('returns false when window is undefined', () => {
+      const originalWindow = global.window;
+      delete (globalThis as { window?: Window }).window;
+
+      const result = isWebHidAvailable();
+
+      expect(result).toBe(false);
+
+      global.window = originalWindow;
```
Comment (ccharly) [inline ui/contexts/hardware-wallets/webConnectionUtils.test.ts]: IMO, this kind of setup/teardown should go on `setupDefaultNavigator` and `restoreNavigator`.

That would reduce those manual teardown a lot!

(also, would avoid test from bleeding out and make other tests to fail)

File: ui/contexts/hardware-wallets/webConnectionUtils.test.ts (hunk 16 above)
Comment (ccharly) [reply ui/contexts/hardware-wallets/webConnectionUtils.test.ts]: Having a `mockHid` and `mockUsb` "globals" to all tests could be an option? 🤔

File: ui/contexts/hardware-wallets/webConnectionUtils.test.ts
Comment (ccharly) [inline ui/contexts/hardware-wallets/webConnectionUtils.test.ts:1]: Some `describe` blocks are no longer matching the actual function names

**Review by ccharly — APPROVED** [review]

**Review by gantunesr — APPROVED** [review]
